### ISSUE TITLE
Test helper - get the original response data

### ISSUE THREAD

**Issue by Mrkisha** [symfony/symfony#23204]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no

So I was watching Adam Wathan and his live TDD Laravel on YT. He did one very interesting thing which resonated with me. When he writes functional tests, he does not assert against returned html (as html can change by front end devs), however Laravel provides this call in test `$this->response->getOriginalContent()->getData()` where you get original data and you can assert against it.

It would be nice if we could have such thing in Symfony. I've tried to get to original response data in test but no success.

**Comment by ogizanagi** [comment]
What are original data? The data gave to the template?

If so, I think I've made something similar in the past, in an application for which we worked with a remote team for frontend. Our functional tests were then only testing that the frontend team will always get expected data in the template for a given request. I made use of a `TemplateParametersDataCollector` and a `TraceableTemplateEngine` registered in test environment, in charge of collecting the `render/renderResponse` second argument (parameters).

Those data where then available from the tests by enabling the profiler before making the request, getting the collector and retrieving data from it once the request was executed. The data were cloned using the var-dumper component, so we were basically testing dumped versions of the parameters, but that fitted well with our needs, as we were only passing DTO objects to the templates. But with the seamless usage of data clones introduced in 3.3, it could be even more powerful.
